src/spox/_attributes.py

```diff
@@ -50,20 +46,44 @@ def value(self) -> T:
         return self._value
 
     def _validate(self):
-        if self._to_onnx("dummy").type != self._attribute_proto_type_int:
-            raise TypeError(
-                f"Unable to instantiate `{type(self).__name__}` with value of type `{type(self.value).__name__}`."
-            )
+        try:
+            type_in_onnx = self._to_onnx_deref("dummy").type
+        except TypeError as e:
+            # TypeError: 'a' has type str, but expected one of: int -- this is raised from within protobuf
+            # when .extend()-ing with the wrong type.
+            if "but expected one of" in str(e):
+                raise self._get_pretty_type_exception() from e
+            else:
+                raise
+
+        if type_in_onnx != self._attribute_proto_type_int:
+            raise self._get_pretty_type_exception()
 
     def _to_onnx(self, key: str) -> AttributeProto:
         if isinstance(self._value, _Ref):
             return self._value._to_onnx(key)
         return self._to_onnx_deref(key)
 
+    @property
+    @abc.abstractmethod
+    def _attribute_proto_type_int(self) -> int:
+        raise NotImplementedError
```

Comment:
I think `NotImplementedError` (but instantiated) is [fitting](https://docs.python.org/3/library/exceptions.html?highlight=notimplementederror#NotImplementedError) here if we want to go with a property. Otherwise, can't we simply have `_attribute_proto_type_int: int` as a field?

Reply:
We get nicer type checks like this:

> 
src/spox/opset/ai/onnx/v17.py:4650: error: Cannot instantiate abstract class "AttrFloat32" with abstract attribute "_attribute_proto_type_int"  [abstract]


![image](https://github.com/Quantco/spox/assets/84503892/beda0534-609d-4f4c-b09e-d0be9f67cbd6)